### PR TITLE
[ValueTracking] Recognize `LShr(UINT_MAX, Y) + 1` as a power-of-two

### PR DESCRIPTION
**Pull request by YanWQ-monad**
There is a missed optimization in
``` llvm
define i8 @known_power_of_two_rust_next_power_of_two(i8 %x, i8 %y) {
  %2 = add i8 %x, -1
  %3 = tail call i8 @llvm.ctlz.i8(i8 %2, i1 true)
  %4 = lshr i8 -1, %3
  %5 = add i8 %4, 1
  %6 = icmp ugt i8 %x, 1
  %p = select i1 %6, i8 %5, i8 1

  %r = urem i8 %y, %p
  ret i8 %r
}
```
which is extracted from the Rust code
``` rust
fn func(x: usize, y: usize) -> usize {
    let z = x.next_power_of_two();
    y % z
}
```
Here `%p` (a.k.a `z`) is semantically a power-of-two, so `y urem p` can be optimized to `y & (p - 1)`. (Alive2 proof: https://alive2.llvm.org/ce/z/H3zooY)

---

It could be generalized to recognizing `LShr(UINT_MAX, Y) + 1` as a power-of-two, which is what this PR does.
Alive2 proof: https://alive2.llvm.org/ce/z/zUPTbc

### CODE REVIEW

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -2173,6 +2174,13 @@ bool isKnownToBeAPowerOfTwo(const Value *V, bool OrZero, unsigned Depth,
         if (OrZero || RHSBits.One.getBoolValue() || LHSBits.One.getBoolValue())
           return true;
     }
+
+    // LShr(UINT_MAX, Y) + 1 is a power of two (if nuw) or zero.
+    if (OrZero || Q.IIQ.hasNoUnsignedWrap(VOBO)) {
+      APInt UIntMax = APInt::getMaxValue(BitWidth);
+      if (match(I, m_Add(m_LShr(m_SpecificInt(UIntMax), m_Value()), m_One())))
```
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:2181]:
```suggestion
      if (match(I, m_Add(m_LShr(m_AllOnes(), m_Value()), m_One())))
```

File: llvm/test/Analysis/ValueTracking/known-power-of-two-urem.ll [hunk 2]
```diff
@@ -387,3 +387,26 @@ for.end:
...
+  %3 = tail call i8 @llvm.ctlz.i8(i8 %2, i1 true)
+  %4 = lshr i8 -1, %3
+  %5 = add i8 %4, 1
+  %6 = icmp ugt i8 %x, 1
+  %p = select i1 %6, i8 %5, i8 1
+  ; Rust's implementation of `%p = next_power_of_two(%x)`
+
+  %r = urem i8 %y, %p
+  ret i8 %r
+}
```
Comment (nikic) [inline llvm/test/Analysis/ValueTracking/known-power-of-two-urem.ll:412]: Could you please also add a more targeted test of just the lshr + add pattern, and then test:

 * The OrZero/nuw logic.
 * Negative tests for wrong constants (-1, 1).

**Review by nikic — COMMENTED** [review]
Nice find.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Nice catch!

BTW, can you please have a look at https://github.com/dtcxzyw/llvm-opt-benchmark/blob/03a9d18338e0213e304f67ea9cb18e4ae95f5d26/bench/regex-rs/optimized/11vfjke4utuj478u.ll#L7147-L7152? I guess we can handle this pattern in `foldCttzCtlz` as well :)

Proof: https://alive2.llvm.org/ce/z/2aHfYa

**Review by nikic — APPROVED** [review]

**Pull request merged**